metaflow/client/core.py

```diff
@@ -1123,6 +1124,153 @@ def _iter_filter(self, x):
         # exclude private data artifacts
         return x.id[0] != "_"
 
+    def _iter_matching_tasks(self, steps, metadata_key, metadata_pattern):
+        """
+        Yield tasks from specified steps matching a foreach path pattern.
+
+        Parameters
+        ----------
+        steps : List[str]
+            List of step names to search for tasks
+        pattern : str
+            Regex pattern to match foreach-indices metadata
+
+        Returns
+        -------
+        Iterator[Task]
+            Tasks matching the foreach path pattern
+        """
+        flow_id, run_id, _, _ = self.path_components
+
+        for step in steps:
+            try:
+                task_pathspecs = self._metaflow.metadata.filter_tasks_by_metadata(
+                    flow_id, run_id, step.id, metadata_key, metadata_pattern
+                )
+            except Exception as e:
+                if e.http_code == 404:
+                    # filter_tasks_by_metadata endpoint does not exist in the version of metadata service
+                    # deployed currently. Raise a more informative error message.
+                    raise MetaflowInternalError(
+                        "The version of metadata service deployed currently does not support filtering tasks by metadata. "
+                        "Upgrade to a newer version of Metadata service to use this feature."
+                    ) from e
+
```

Comment:
this code block can be within the service implementation and not in the client. the client shouldn't have to bother about the actual implementation semantics of metadata.

Reply:
If the endpoint does not exist in the metadata service then how do you even return that error message?

Reply:
this try-except block can be in service.py rather than in the client.